### ISSUE TITLE
have an export_translation that tells godot the exported string needs translation

### ISSUE THREAD

**Issue by Dracks** [godotengine/godot-proposals#10139]
### Describe the project you are working on

I'm working on a project that is a clone of advance wars, and I was trying to cleanup/improve our translation workflows

### Describe the problem or limitation you are having in your project

Currently we have some custom scripts for scenes, that contains exported variables, that needs to be translated, and we need to manually do the call to the translation, like: 
```
@export var title: String = "Menu":
	set(value):
		($Label as Label).text = tr(value)
		title = tr(value)
``` 

Also, this means the automatic pot generation don't detect this texts.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I will like to have some export_translation that automatically can translate the string without custom code, Also will be nice that godot can pick the different texts used in this exported values to generate the base.pot file

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

use the new flag in the string, to rewrite my previous code:
```
@export var title: String = "Menu":
	set(value):
		($Label as Label).text = value
		title = value

```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

more or less yes, for the automatic pot generation is more problematic

### Is there a reason why this should be core and not an add-on in the asset library?

a library/plugin cannot control this as far as I know

**Comment by RedMser** [comment]
Your proposal code is missing use of the new annotation.

As for workarounds, you could always prefix those fields with a custom group or a name prefix like `tr_title`, loop through `get_property_list` in an external script and have that do the translation and detect language changes.

**Comment by Dracks** [comment]
> Your proposal code is missing use of the new annotation.
Sorry, Changed now!

> As for workarounds, you could always prefix those fields with a custom group or a name prefix like `tr_title`, loop through `get_property_list` in an external script and have that do the translation and detect language changes.

I  don't see the benefit of your workarround vs our current one, maybe I'm missing something?

**Comment by RedMser** [comment]
> I don't see the benefit of your workarround vs our current one, maybe I'm missing something?

Automatic POT generator can be extended with a plugin to detect which fields are marked as translatable just by getting list of script properties.

Also your current solution doesn't handle the language change notification, so you'd need to restart the game or reload the entire scene to change language (or handle the notification as boilerplate in each script).

**Comment by Dracks** [comment]
Ah, I didn't know that the plugins can modify the POT generator. 

Then your workarround will be to have some kind of autoload? that detects it? Also it can receive signals? 

About the change of language, you mean when we are inside the game, and we change the language there right? 

Thanks a lot for your input!

**Comment by RedMser** [comment]
> I didn't know that the plugins can modify the POT generator.

See [EditorTranslationParserPlugin](https://docs.godotengine.org/en/stable/classes/class_editortranslationparserplugin.html) and [EditorPlugin.add_translation_parser_plugin](https://docs.godotengine.org/en/stable/classes/class_editorplugin.html#class-editorplugin-method-add-translation-parser-plugin)

> Then your workarround will be to have some kind of autoload? that detects it? Also it can receive signals?
> 
> About the change of language, you mean when we are inside the game, and we change the language there right?

Yes I haven't tried it myself like this, but it seems very possible to do it. Change of language is indeed like you said, if you're in a settings menu and change the language dropdown. You need to handle [NOTIFICATION_TRANSLATION_CHANGED](https://docs.godotengine.org/en/latest/classes/class_node.html#class-node-constant-notification-translation-changed) as shown in the docs.

---

By the way I'm not saying this proposal would not be helpful. Just providing a workaround so everyone knows there is alternative ways to approach the problem, and maybe the core team decides this is e.g. something that a custom annotation should handle instead of a core builtin (see https://github.com/godotengine/godot-proposals/issues/1316 )

**Comment by Dracks** [comment]
I never said that you were discarting my "proposal", at the end is only a proposal.

I'm learning a lot in this ticket! (tecnically I already added a link to this issue in our code) 

I already found the editorTranslationParserPlugin. But I don't expect to have more time to invest on this for the future weeks. Let's see... Still super thanks!

yeah, maybe a custom annotation will work also, I don't know about them yet. (They are also a proposal, right?)

Thanks!

**Comment by KoBeWi** [comment]
Unfortunately EditorTranslationParserPlugin *overwrites* the built-in plugins, if they handle the same extension. Thus it's not possible to extend current parser, you can only replace it, or add a parser for custom file, which seems to be what the plugin is currently intended for..

The build-in parsers are rather dumb and will use hard-coded list of properties to determine what to translate. You can abuse that by naming your translateable strings `text` and similar. Though it's not documented, so you have to lookup the source, also the list is limited and not flexible. It's currently the only workaround though (other than manually adding strings to POT, which is what I do currently).

The solution proposed here is what I thought of at first and is probably the easiest one to use.
Others include allowing multiple parser plugins to handle the same extension, so you can implement additional parsing logic, or exposing PackedSceneEditorTranslationParserPlugin so users can extend it and add custom parsing logic.

**Comment by clarence112** [comment]
I've dug through the code. It seems like a current workaround is that any string variable who's name ends with `_text` will be translated.  
So `@export anynameyouwant_text:String` will be translatable

**Comment by bluebirdsan** [comment]
I think an `@export_translation` would be a good addition.

While adding `_text` does work, it is superfluous, because the `String` type is already text. `_text` increases the variable name, increasing the clutter in the code.

Plus adding `_text` is not obvious (I found it while searching for bug reports/proposals, because I thought this was a bug).
I think searching through `@export_*` would be one of the first things to try, because it would follow an existing way to modify exported variables.

**Comment by KoBeWi** [comment]
btw
```GDscript
@export var title: String = "Menu":
	set(value):
		($Label as Label).text = tr(value)
		title = tr(value)
```
You don't need `tr(value)` for Label, as it will translate the string automatically.
Also you shouldn't assign translated string to `title`, only use `tr()` for displayed text (e.g. when you draw it with `draw_string()`).

**Comment by bluebirdsan** [comment]
I started implementing this, but have a question.

Would it be best if ending String variable names with `_text` would include them in the translation template or not?

I think it would be better that doing that wouldn't work anymore and you'd have to use the `@export_translation` annotation as it would make it very explicit. Otherwise there would be confusion where some String variables end with `_text` and others have the annotation, but both do the same thing.

I'd like to know other people's arguments on this.

**Comment by clarence112** [comment]
> I think it would be better that doing that wouldn't work anymore

That change could be a pretty significant compat breakage.

Personally I haven't needed to rely on it but who knows about anyone else.

**Comment by bluebirdsan** [comment]
That's true. Probably a good idea is to have both ways to add Strings to translation and remove the `_text` way in the eventual Godot 5.0 release.

**Comment by KoBeWi** [comment]
Technically it was never documented, so we don't need to keep this behavior.

**Comment by Gonderage** [comment]
For what it's worth, there already exists `PROPERTY_HINT_LOCALIZABLE_STRING` that strangely only exists for `TYPE_DICTIONARY` despite having `STRING` in the name, and there's already precedent for `PROPERTY_HINT_*` constants working for multiple types, so I could forsee myslef using this a lot for `@export_custom`, and would save having to add a new constant.
